feat(forms): Add `FormArrayDirective`

JeanMeche opened the pull request:
The `FormArrayDirective` will allow to have a `FormArray` as a top-level form object.

* `NgControlStatusGroup` directive will be applied to the `FormArrayDirective`
* `NgForm` will still create a `FormGroup`

Usage : 

```
@Component({
  selector: 'form-array-comp',
  template: `
    <form [formArray]="form">
      @for(_ of controls; track $index) {
        <input [formControlName]="$index">
      }
    </form>`,
})
class FormArrayComp {
  controls = [new FormControl('fish'), new FormControl('cat'), new FormControl('dog')];
  form = new FormArray(this.controls);
}
```


Fixes https://github.com/angular/angular/issues/30264

BREAKING CHANGE: The `formDirective` getter on `FormGroupName` and `FormArrayName` now return a `AbstractFormDirective`

AndrewKushnir reviewed (commented):
@JeanMeche I had a quick look and the changes look good. I want to ask if you could split the commit into 2:

- first `refactor` commit performs a refactoring and extracts out the common logic into the abstract form directive class
- the second `feat` commit that adds a new directive

That should help to make the change more incremental and help with more detailed code review.

JeanMeche committed:
refactor(forms): extract shared logic from `FormGroupDirective`

Ahead of the implementation of `FormArrayDirective`, extract the shared logic into an abstract class.

JeanMeche committed:
feat(forms): Add FormArrayDirective

The `FormArrayDirective` will allow to have a `FormArray` as a top-level form object.

* `NgControlStatusGroup` directive will be applied to the `FormArrayDirective`
* `NgForm` will still create a `FormGroup`

Fixes angular#30264

BREAKING CHANGE: This new directive will conflict with existing FormArray directives or formArray inputs on the same element.

dylhunn reviewed (approved):
Thanks for splitting up the change in this way! Let's merge this *after* the RC, so we can get it on the 18.2 track, and we'll have plenty of time to receive feedback from GDEs and other early users.

AndrewKushnir commented on packages/forms/src/directives/reactive_directives/form_group_name.ts:
```diff
@@ -127,12 +127,12 @@ export const formArrayNameProvider: any = {
  * It accepts the string name of the nested `FormArray` you want to link, and
  * will look for a `FormArray` registered with that name in the parent
- * `FormGroup` instance you passed into `FormGroupDirective`.
+ * `FormGroup`/`FormArray` instance you passed into `FormGroupDirective`/`FormGroupArrat`.
```
Typo: `FormGroupArrat` -> `FormGroupArray`

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -5812,6 +5825,132 @@ describe('reactive forms integration tests', () => {
       }).not.toThrow();
     });
+
+    describe('formArray support', () => {
```
I'm curious if we may want to add a check that `[formArray]` and `[formGroup]` should not be present on the same element? Can we have valid use-cases where it might be needed (e.g. they bind to different data structures)?

AndrewKushnir commented on packages/forms/src/directives/reactive_directives/form_array_directive.ts:
```diff
@@ -0,0 +1,71 @@
+ * @see {@link AbstractControl}
+ *
+ * @usageNotes
+ * ### Register Form Group
```
"Form Group" -> "Form Array" (also an `@example` link looks potentially incorrect)

JeanMeche replied:
Non-inline examples don't work with ADEV (at least ATM, I'll remove it).

AndrewKushnir commented on packages/forms/src/directives/reactive_directives/form_group_directive.ts:
```diff
@@ -6,49 +6,13 @@
-import {FormArrayName, FormGroupName} from './form_group_name';
-import {FormResetEvent, FormSubmittedEvent} from '../../model/abstract_model';
+import {AbstractFormDirective} from './abstract_form.directive';
+import {AbstractControl} from '../../forms';
```
I think we can avoid circular dep here if we import `AbstractControl` from a more specific location (e.g. '../../forms/model/abstract_model').

AndrewKushnir reviewed (commented):
The change looks good, thanks for splitting the change into 2 commits 👍 

I've left a few comments, but a bigger question that I have is about potential collisions with existing `[formArray]` usages in templates:

- as a custom directive
- as an input to a component

I've checked internal codebase and there are ~75 instances of the `[formArray]` in templates. Adding a new directive to the `ReactiveFormsModule` will cause conflicts with those locations.

I think we'd need to do some research to find a way to avoid those conflicts or do some cleanup internally. cc @dylhunn to get your opinion on this as well.

JeanMeche commented:
I'm moving this to major / breaking change as it was discovered that there were already usages of a formArray directive (or formArray as input) inside Google.

JeanMeche commented:
Converting back to draft until we enter the breaking change window.

thePunderWoman reviewed (approved):
LGTM!

reviewed-for: fw-general, public-api

mmalerba reviewed (approved):
reviewed-for: public-api
